File: backend/app/models/member.py
"""
會員模型
"""
from sqlalchemy import Column, String, Boolean, Date, DateTime, Text, BigInteger, Integer
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.models.base import Base


class Member(Base):
    """會員表"""

    __tablename__ = "members"

    id = Column(BigInteger, primary_key=True, autoincrement=True)

    # LINE 相關資訊
    line_uid = Column(String(100), unique=True, index=True, comment="LINE UID")
    line_channel_id = Column(String(100), index=True, comment="LINE 官方 Channel ID，對應 line_channels.channel_id")
    line_avatar = Column(String(500), comment="LINE 會員頭像 CDN URL（儲存 LINE 提供的完整 URL，如 https://profile.line-scdn.net/xxxxx），若無頭像或 URL 失效則顯示預設頭像。URL 來源：會員加入時從 LINE Profile API 取得，儲存後不定期更新。前端顯示時直接載入此 URL")
    line_display_name = Column(String(100), comment="LINE 顯示名稱")

    # Facebook 相關資訊
    fb_customer_id = Column(String(100), unique=True, index=True, comment="Facebook Customer ID，透過 Facebook OAuth 登入時取得")
    fb_avatar = Column(String(500), comment="Facebook 會員頭像 URL")
    fb_customer_name = Column(String(100), comment="Facebook 顯示名稱")

    # Webchat 相關資訊
    webchat_uid = Column(String(100), unique=True, index=True, comment="Webchat 訪客 ID，系統自動生成或透過 OAuth 關聯取得")
    webchat_avatar = Column(String(500), comment="Webchat 會員頭像 URL")
    webchat_name = Column(String(100), comment="Webchat 顯示名稱")
    webchat_site_id = Column(String(50), index=True, comment="Webchat widget 嵌入站點英文代號（例：starbit-ryan）")
    webchat_site_name = Column(String(100), comment="Webchat widget 嵌入站點顯示名（例：思偉達飯店｜雷恩館）")

    # 基本資訊
    name = Column(String(32), comment="會員姓名（統一單欄位）")
    gender = Column(String(1), nullable=False, default="0", server_default="0", comment="性別：0=不透漏/1=男/2=女")
    birthday = Column(Date, comment="生日")
    email = Column(String(255), unique=True, index=True, comment="電子信箱")
    phone = Column(String(20), index=True, comment="手機號碼")
    id_number = Column(String(20), unique=True, index=True, comment="身分證字號")
    passport_number = Column(String(50), comment="護照號碼")
    residence = Column(String(100), comment="居住地")
    address_detail = Column(String(255), comment="詳細地址")

    # 系統資訊
    join_source = Column(
        String(20),
        nullable=False,
        default="LINE",
        comment="加入來源：LINE/CRM/PMS/ERP/系統/Webchat",
        index=True,
    )
    receive_notification = Column(Boolean, default=True, comment="是否接收優惠通知")
    gpt_enabled = Column(Boolean, default=True, comment="是否啟用 GPT 自動回應")
    # 訪客（未加入會員的 webchat widget 使用者）
    is_guest = Column(
        Boolean,
        nullable=False,
        default=False,
        server_default="0",
        comment="是否為匿名 webchat 訪客（未填寫資料）",
        index=True,
    )
    guest_seq = Column(
        Integer,
        unique=True,
        nullable=True,
        comment="訪客流水號（顯示為 訪客{guest_seq:06d}）",
    )
    human_override_until = Column(DateTime, nullable=True, comment="人工接管到期時間（UTC），非 NULL 且未過期時抑制所有自動回應")
    internal_note = Column(Text, comment="內部備註")
    last_interaction_at = Column(DateTime, index=True, comment="最後互動時間")
    is_following = Column(
        Boolean,
        nullable=True,
        default=True,
        server_default="1",
        comment="是否正在關注 LINE OA（從 line_friends 同步）",
    )

    created_at = Column(DateTime, server_default=func.now(), comment="建立時間")
    updated_at = Column(DateTime, onupdate=func.now(), comment="更新時間")

    # 關聯關係
    member_tags = relationship(
        "MemberTag", back_populates="member", cascade="all, delete-orphan"
    )
    member_interaction_tags = relationship(
        "MemberInteractionTag", back_populates="member", cascade="all, delete-orphan"
    )
    consumption_records = relationship(
        "ConsumptionRecord", back_populates="member", cascade="all, delete-orphan"
    )
    pms_integrations = relationship(
        "PMSIntegration", back_populates="member", cascade="all, delete-orphan"
    )
    message_deliveries = relationship(
        "MessageDelivery", back_populates="member", cascade="all, delete-orphan"
    )

    @property
    def message_recipients(self):
        """向後兼容舊屬性名稱"""
        return self.message_deliveries

    @message_recipients.setter
    def message_recipients(self, value):
        self.message_deliveries = value

File: frontend/src/types/member.ts
import type { MemberSourceType } from './channel';

/**
 * 標籤資訊
 * 包含標籤的完整資訊
 */
export interface TagInfo {
  id: number;
  name: string;
  type: 'member' | 'interaction';
}

/**
 * 基础会员信息
 * 包含所有会员的核心字段
 */
export interface Member {
  id: string;
  username: string;  // LINE user name
  realName: string;
  tags: string[];
  memberTags?: string[];      // 會員標籤（字串陣列，向後相容）
  interactionTags?: string[]; // 互動標籤（字串陣列，向後相容）
  tagDetails?: TagInfo[];     // 標籤詳細資訊
  phone: string;
  email: string;
  gender?: string;            // 性別：0=不透露/1=男/2=女 或 undisclosed/male/female
  birthday?: string;          // 生日 (ISO 日期格式 YYYY-MM-DD)
  createTime: string;
  lastChatTime: string;
  // LINE 渠道
  lineUid?: string;
  lineAvatar?: string;
  line_display_name?: string;
  channel_id?: string;        // LINE channel ID
  // Facebook 渠道
  fb_customer_id?: string;
  fb_customer_name?: string;
  fb_avatar?: string;
  // Webchat 渠道
  webchat_uid?: string;
  webchat_name?: string;
  webchat_avatar?: string;
  // 其他
  join_source?: MemberSourceType; // 加入來源：LINE/CRM/PMS/ERP/系統 - 使用統一類型
  id_number?: string;         // 身分證字號
  residence?: string;         // 居住地
  passport_number?: string;   // 護照號碼
  internal_note?: string;     // 會員備註
  gpt_enabled?: boolean;      // GPT 自動回覆模式 (true=自動, false=手動)
}

/**
 * 渠道類型
 */
export type ChannelType = 'LINE' | 'Facebook' | 'Webchat';

/**
 * 會員管理頁表格顯示用型別
 * LINE 和 FB 分開取、直接顯示
 */
export interface DisplayMember {
  id: string;                    // 唯一 ID (如 "line-123", "fb-456")
  odooMemberId: number | null;   // 本地 DB 的 member.id（FB 可能為 null）
  channel: ChannelType;          // 渠道類型
  channelUid: string;            // 渠道 UID (line_uid 或 customer_id)
  displayName: string;           // 顯示名稱（渠道名稱，如 LINE 暱稱）
  realName: string | null;       // 真實姓名（從會員資料）
  avatar: string | null;         // 頭像 URL
  email: string | null;          // Email
  phone: string | null;          // 電話
  createTime: string | null;     // 建立時間
  lastChatTime: string | null;   // 最後聊天時間
  tags: string[];                // 標籤列表
  // 未回覆狀態
  isUnanswered: boolean;         // 是否有未回覆的訊息
  unansweredSince: string | null; // 未回覆訊息的時間戳
  // 渠道帳號名稱
  channelName: string | null;    // 渠道帳號名稱 (e.g., "下班解憂所")
}

/**
 * 会员列表项（用于显示）
 * 可以根据需要添加额外的 UI 状态
 */
export interface MemberListItem extends Member {
  selected?: boolean;
  expanded?: boolean;
}

/**
 * 会员表单数据
 * 用于创建或编辑会员
 */
export interface MemberFormData {
  username?: string;
  realName?: string;
  tags?: string[];
  phone?: string;
  email?: string;
  note?: string;
}

/**
 * 类型守卫：检查是否为有效的会员对象
 */
export function isMember(obj: any): obj is Member {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    typeof obj.id === 'string' &&
    typeof obj.username === 'string' &&
    typeof obj.realName === 'string' &&
    Array.isArray(obj.tags) &&
    typeof obj.phone === 'string' &&
    typeof obj.email === 'string' &&
    typeof obj.createTime === 'string' &&
    typeof obj.lastChatTime === 'string'
  );
}

/**
 * 工具函数：创建空的会员对象
 * 用于初始化表单或测试
 */
export function createEmptyMember(): Member {
  return {
    id: '',
    username: '',
    realName: '',
    tags: [],
    phone: '',
    email: '',
    createTime: new Date().toISOString(),
    lastChatTime: new Date().toISOString(),
    lineUid: '',
    lineAvatar: '',
  };
}
